src/main/java/org/opensearch/neuralsearch/processor/optimization/InferenceFilter.java

```diff
@@ -0,0 +1,233 @@
+     * Map values: Processed recursively using this method.
+     * List values: Processed using filterListValue.
+     * Primitive values: Directly compared using filterInferenceValue.
+     *
+     * @param existingSourceAndMetadataMap The metadata map of the existing document.
+     * @param sourceAndMetadataMap The metadata map of the new document.
+     * @param processMap The current map being processed.
+     *
+     * @return A filtered map containing only elements that require new embeddings.
+     *
+     */
+    public Map<String, Object> filter(
+        Map<String, Object> existingSourceAndMetadataMap,
+        Map<String, Object> sourceAndMetadataMap,
+        Map<String, Object> processMap
+    ) {
+        return filter(existingSourceAndMetadataMap, sourceAndMetadataMap, processMap, "");
+    }
+
+    /**
+     * Helper method for filter
+     * @param existingSourceAndMetadataMap The metadata map of the existing document.
+     * @param sourceAndMetadataMap The metadata map of the new document.
+     * @param processMap The current map being processed.
+     * @param traversedPath The dot-notation path of the previously traversed elements.
+     * e.g:
+     * In a map structured like:
+     *       level1
+     *         level2
+     *            level3
+     * traversedPath would be dot-separated string of level1.level2.level3
+     * @return A filtered map containing only elements that require new embeddings
+     */
+    private Map<String, Object> filter(
+        Map<String, Object> existingSourceAndMetadataMap,
+        Map<String, Object> sourceAndMetadataMap,
+        Object processMap,
+        String traversedPath
+    ) {
+        if (processMap instanceof Map == false) {
+            throw new IllegalArgumentException("processMap needs to be an instanceof Map");
+        }
+        Map<String, Object> filteredProcessMap = new HashMap<>();
+        Map<String, Object> castedProcessMap = ProcessorUtils.castToMap(processMap);
+        for (Map.Entry<?, ?> entry : castedProcessMap.entrySet()) {
+            if ((entry.getKey() instanceof String) == false) {
+                throw new IllegalArgumentException("key for processMap must be a string");
+            }
+            String key = (String) entry.getKey();
+            Object value = entry.getValue();
+            String currentPath = traversedPath.isEmpty() ? key : traversedPath + "." + key;
+            if (value instanceof Map<?, ?>) {
+                Map<String, Object> filteredInnerMap = filter(existingSourceAndMetadataMap, sourceAndMetadataMap, value, currentPath);
+                filteredProcessMap.put(key, filteredInnerMap.isEmpty() ? null : filteredInnerMap);
+            } else if (value instanceof List) {
+                List<Object> processedList = filterListValue(
+                    currentPath,
+                    (List<Object>) value,
```

Comment:
This is not safe as well. If you know that value will be always List<Object>, let's use
```
@SuppressWarnings("unchecked")
    public static List<Object> castToObjectList(List<?> obj) {
        return (List<Object>) obj;
    }
```